### DISCUSSION TITLE
How do I ensure that materials are one-sided or two-sided?

### DISCUSSION THREAD

**Discussion by hjeldin** [google/model-viewer#1867] [Q&A]
### Description

Models are loaded with material.side = THREE.DoubleSided, which causes rendering issues with non-opaque materials.

![Schermata 2019-09-03 alle 16 47 03](https://user-images.githubusercontent.com/922567/64183652-7f5a7280-ce6a-11e9-8728-92eb775c8f70.png)

I don't recall seeing any API to access the THREE.js layer, so i'm kinda out of alternatives, but setting THREE.FrontSide or THREE.BackSide to the side property of the material should suffice.

#### Live Demo
https://glitch.com/edit/#!/granite-lemur

### Browser Affected

- [x] Chrome
- [ ] Edge
- [ ] Firefox
- [ ] Helios
- [ ] IE
- [x] Safari

### OS

- [x] Android
- [x] iOS
- [ ] Linux
- [x] MacOS
- [ ] Windows

### Versions
- model-viewer: v0.6.0
- three.js: rXX

**Comment by cdata** [comment]
Hey, thanks for checking in with us before giving up 👍 I did some reading and took a stab at tweaking your model for better results. Let me know if this addresses your problem!

Scanning the glTF spec, it looks like [materials support being configured as double sided](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#double-sided).

Scanning the `GLTFLoader` source code, it also looks like [Three.js respects this property](https://github.com/mrdoob/three.js/blob/f3205d7cc2b1478c8e80fd06490d552b2d44eb6f/examples/js/loaders/GLTFLoader.js#L2275-L2279
).

Popping your model into Blender 2.8, I was able to turn on backface culling:

![bfcull](https://user-images.githubusercontent.com/240083/64186958-8fe7fa00-ce24-11e9-9fdc-be8e89728fe3.png)

...leading to this improved result:

![image](https://user-images.githubusercontent.com/240083/64187046-bad24e00-ce24-11e9-987e-c001bd390904.png)

Here is my tweaked version of your demo model: [export-fixed.glb.zip](https://github.com/GoogleWebComponents/model-viewer/files/3570242/export-fixed.glb.zip)

**Comment by hjeldin** [comment]
Yep, it just occurred to me that we're using an automated pipeline to export the models. One of the stages of the pipeline is actually obj2gltf, which automatically sets DoubleSided on the material properties if the material is transparent.

Thanks for your help!

Should problems like these rise again, have you considered exposing some parts of the rendering layer to the model-viewer api?

**Comment by cdata** [comment]
Yes! We have speculated about exposing the scene graph via a "render worklet" of some kind (previous public discussion [here](https://github.com/GoogleWebComponents/model-viewer/issues/186#issuecomment-439303818) and [here](https://github.com/GoogleWebComponents/model-viewer/issues/685#issuecomment-516614906)). This wouldn't give you direct access to the Three.js API, but it would give you a lot of flexibility to tweak the models at runtime.

Since this has come up a few times, I have now filed https://github.com/GoogleWebComponents/model-viewer/issues/738 to cover our current ideas on the matter. Your feedback would be welcome there 👍
